Sources/TensorFlow/Operators/LinearAlgebra.swift

```diff
@@ -106,6 +106,22 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
     }
 }
 
+/// Computes the trace of an input tensor.
+/// Returns the sum along the main diagonal of [batched] matrix.
+///
+/// The input has to be [batched] square matrix.
+/// When the input is a tensor of shape `[..., M, M]` then the output is a tensor
+/// with dimensions `[...]`
+///
+/// - Parameter matrix: A tensor of shape `[..., M, M]`
+@inlinable
+@differentiable(wrt: matrix where T: TensorFlowFloatingPoint)
+public func trace<T: TensorFlowNumeric>(_ matrix: Tensor<T>) -> Tensor<T> {
+    precondition(matrix.rank >= 2, "The tensor must have at least rank 2.")
+    precondition(matrix.shape[matrix.rank - 1] == matrix.shape[matrix.rank - 2], "Squared [batched] matrix expected")
```

Comment:
Python TensorFlow doesn't seem to have a "square matrix" requirement:
```python
import tensorflow as tf
print(tf.__version__) # 2.0.0-beta1
x = tf.constant([[1, 2], [3, 4], [5, 6]])
print(x.shape) # (3, 2), not square
print(tf.linalg.trace(x)) # tf.Tensor(5, shape=(), dtype=int32)
```
[The documentation for `tf.linalg.trace`](https://www.tensorflow.org/api_docs/python/tf/linalg/trace) suggests the same: inputs with shape `[..., M, N]` are supported.

Did you add this precondition for a reason? I think lifting the precondition and matching TensorFlow's behavior would be good, as a rule of thumb.

Reply:
From, https://en.wikipedia.org/wiki/Trace_(linear_algebra)

> **The trace is only defined for a square matrix (n × n).**

TensorFlow is wrong :)

Reply:
This brings us to the question of whether we should match the behavior of Python TensorFlow (for exact reproducibility, which is an important use case and may affect model convergence), or pursue different, more sensible/correct behavior.

In this case, adding an additional precondition seems "safe", because it doesn't lead to silently different computation results compared with Python TensorFlow. There aren't yet users of `trace` that want it to work for non-square matrices. So adding the precondition sounds good to me!

---

Another scenario: I recall a previous discussion with @eaplatanios about using a different (more mathematically correct?) derivative implementation for a particular function than Python TensorFlow. We decided against that change because it does lead to different computation results compared with Python TensorFlow, breaking exact reproducibility.

Reply:
Apparently, the numpy does the same mistake... or a feature... Should I remove that condition then?

Reply:
> Apparently, the numpy does the same mistake... or a feature... Should I remove that condition then?

Ah. In that case, could you please remove the precondition and update doc comments?
(Mentions of "square matrix" should be removed, and `M, M` should be replaced with `M, N`.)

Reply:
done.

Reply:
Thanks! I made minor edits in https://github.com/tensorflow/swift-apis/pull/586/commits/c1407b673a5e89a3078c57f27facf9bb1d7b713d, hope you don't mind.